### PR TITLE
[ARM] arm_acle.h add Coprocessor Instrinsics

### PR DESCRIPTION
**Pull request by hstk30-hw**
https://github.com/llvm/llvm-project/issues/75424 

Add Coprocessor Instrinsics

### CODE REVIEW

File: clang/lib/Headers/arm_acle.h [hunk 1]
```diff
@@ -752,10 +752,57 @@ __arm_st64bv0(void *__addr, data512_t __value) {
 #define __arm_mte_ptrdiff(__ptra, __ptrb) __builtin_arm_subp(__ptra, __ptrb)
 
 /* Memory Operations Intrinsics */
-#define __arm_mops_memset_tag(__tagged_address, __value, __size)    \
+#define __arm_mops_memset_tag(__tagged_address, __value, __size)               \
   __builtin_arm_mops_memset_tag(__tagged_address, __value, __size)
 #endif
 
+/* Coprocessor Intrinsics */
+#if (!__thumb__ || __thumb2__) && __ARM_ARCH >= 4
```
Comment (compnerd) [inline clang/lib/Headers/arm_acle.h:760]: I think that if we are wrapping the version in to the same clause I would rather that we split this condition into:

```suggestion
#if (!__thumb__ || __thumb2__)
#if __ARM_ARCH >= 4
```

That makes all the versions uniform and makes it more obvious that the ARM or Thumb2 mode applies to all versions.

**Review by compnerd — APPROVED** [review]

File: clang/lib/Headers/arm_acle.h [hunk 2]
```diff
@@ -752,10 +752,57 @@ __arm_st64bv0(void *__addr, data512_t __value) {
 #define __arm_mte_ptrdiff(__ptra, __ptrb) __builtin_arm_subp(__ptra, __ptrb)
 
 /* Memory Operations Intrinsics */
-#define __arm_mops_memset_tag(__tagged_address, __value, __size)    \
+#define __arm_mops_memset_tag(__tagged_address, __value, __size)               \
```
Comment (vhscampos) [inline clang/lib/Headers/arm_acle.h:755]: This whitespace change looks unrelated. Can you please put it in a different pull request?

**Review by vhscampos — CHANGES_REQUESTED** [review]

**Comment by vhscampos** [comment]
There are some typos in the commit message and in the PR title.

**Comment by davemgreen** [comment]
It looks like there is a downstream implementation of this that was never upstreamed. Perhaps someone can fish it out for you to show how it looked? It might be using the wrong predefined macro, but does have some tests.

**Comment by davemgreen** [comment]
Let me try and get the downstream version, you might be able to pick up some things from it. A test at least should probably be present.

**Comment by davemgreen** [comment]
This is the downstream code we have: https://gist.github.com/davemgreen/e7ade833274a60e975e67a66eda7cb44
Note that the __ARM_TARGET_COPROC_XYZ macros are probably wrong. They should be __ARM_FEATURE_COPROC bitfield macros according to the ACLE.

Can you make use of some of that? It would be good to add the macro definition at the same time as the intrinsics (they can be used to control when the intrinsics are available), and the test should be useful for checking they are available at the right times.

**Comment by davemgreen** [comment]
Thanks for doing this.
I think that __ARM_FEATURE_COPROC should be a bitfield, as defined in https://arm-software.github.io/acle/main/acle.html#coprocessor-intrinsics. That would remove the need for the other macros.

File: clang/test/CodeGen/arm-acle-coproc.c [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+  // CHECK-V6M-NOT: __builtin_arm_cdp2
+  // CHECK-V7: __builtin_arm_cdp2
+  // CHECK-V8-NOT: __builtin_arm_cdp2
+  // CHECK-V8-BASE-NOT: __builtin_arm_cdp2
+  // CHECK-V8-MAIN: __builtin_arm_cdp2
+}
+
+void ldc(int i) {
+  __arm_ldc(1, 2, &i);
+  // CHECK-V4: __builtin_arm_ldc
```
Comment (hstk30-hw) [inline clang/test/CodeGen/arm-acle-coproc.c:78]: plz check it, maybe a typo for V4-THUMB @davemgreen

File: clang/test/CodeGen/arm-acle-coproc.c [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+  // CHECK-V4-NOT: __builtin_arm_ldc
```
Comment (hstk30-hw) [inline clang/test/CodeGen/arm-acle-coproc.c:79]: plz check it, maybe a typo for V4-THUMB @davemgreen

**Comment by hstk30-hw** [comment]
Use bitfield still a mess up because the different arm version have different  Instrinsics available even in the same bit group.
I will try it, hope it readable.

File: clang/lib/Basic/Targets/ARM.cpp [hunk 5]
```diff
@@ -836,6 +837,70 @@ void ARMTargetInfo::getTargetDefines(const LangOptions &Opts,
...
+    break;
+  case llvm::ARM::ArchKind::ARMV8MMainline:
+    FeatureCoprocBF = FEATURE_COPROC_B1 | FEATURE_COPROC_B2 |
+                      FEATURE_COPROC_B3 | FEATURE_COPROC_B4;
+    break;
+  case llvm::ARM::ArchKind::ARMV9A:
+  case llvm::ARM::ArchKind::ARMV9_1A:
+  case llvm::ARM::ArchKind::ARMV9_2A:
+  case llvm::ARM::ArchKind::ARMV9_3A:
+  case llvm::ARM::ArchKind::ARMV9_4A:
```
Comment (davemgreen) [inline clang/lib/Basic/Targets/ARM.cpp:896]: There is a ARMV9_5A now too. I think I would expect these to be the same as ARMV8.
Is this switch statement exhaustive? Could the default case be made the same as ARMV8 so we don't need to extend it every time an architecture is added?

File: clang/lib/Basic/Targets/ARM.cpp [hunk 6]
```diff
@@ -836,6 +837,70 @@ void ARMTargetInfo::getTargetDefines(const LangOptions &Opts,
...
+  case llvm::ARM::ArchKind::ARMV8_4A:
+  case llvm::ARM::ArchKind::ARMV8_5A:
+  case llvm::ARM::ArchKind::ARMV8_6A:
+  case llvm::ARM::ArchKind::ARMV8_7A:
+  case llvm::ARM::ArchKind::ARMV8_8A:
+  case llvm::ARM::ArchKind::ARMV8_9A:
+    // Filter __arm_cdp, __arm_ldcl, __arm_stcl in arm_acle.h
+    FeatureCoprocBF = FEATURE_COPROC_B1 | FEATURE_COPROC_B3;
+    break;
+  case llvm::ARM::ArchKind::ARMV8MMainline:
```
Comment (davemgreen) [inline clang/lib/Basic/Targets/ARM.cpp:888]: Add ARMV8_1MMainline too.

File: clang/lib/Headers/arm_acle.h [hunk 7]
```diff
@@ -756,6 +756,58 @@ __arm_st64bv0(void *__addr, data512_t __value) {
   __builtin_arm_mops_memset_tag(__tagged_address, __value, __size)
 #endif
 
+/* Coprocessor Intrinsics */
+#if defined(__ARM_FEATURE_COPROC)
+
+#if (__ARM_FEATURE_COPROC & 0x1)
+
+#if (__ARM_ARCH != 8)
```
Comment (davemgreen) [inline clang/lib/Headers/arm_acle.h:764]: Could this be < 8?
This doesn't apply to 8-m.main, right? The test looks OK.

**Review by davemgreen — COMMENTED** [review]
Thanks. This is looking good to me. I just have a few comments about different architecture revisions.

File: clang/lib/Basic/Targets/ARM.cpp (hunk 5 above)
Comment (hstk30-hw) [reply clang/lib/Basic/Targets/ARM.cpp:898]: We need defined ARMV9_5A in ARMTargetParser.def first.

Comment (hstk30-hw) [reply clang/lib/Basic/Targets/ARM.cpp:898]: In this https://gist.github.com/davemgreen/e7ade833274a60e975e67a66eda7cb44, not have  test cases on ARMV9. According to the code logic, ARMV9 seem support all Coprocessor Instrinsics. This is different from ARMV8, I'm not sure, so the default case is 0 for now.

Comment (davemgreen) [reply clang/lib/Basic/Targets/ARM.cpp:898]: Oh right, ARMV9_5A is AArch64 only. That's OK then.
I would expect the other ArmV9-A cases to be the same as ArmV8-A for AArch32, and wouldn't have expected a change in coprocessor instructions.
The reference manual is at https://developer.arm.com/documentation/ddi0487/ja/?lang=en and doesn't seem to mention cdp.

**Comment by davemgreen** [comment]
If you can make armv9-a work the same as armv8-a and add some tests for it then this LGTM

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Comment by hstk30-hw** [comment]
> There are some typos in the commit message and in the PR title.

@vhscampos Check again. Fixed the typos.

**Review by vhscampos — APPROVED** [review]
LGTM. Thanks
